src/MgvOfferTaking.sol

```diff
@@ -211,108 +227,96 @@ abstract contract MgvOfferTaking is MgvHasOffers {
 
-        (gasused, makerData, mgvData) = execute(pair, mor, sor);
+      /* Keep cached copy of current `sor` values to restore them later to send to posthook. */
+      uint takerWants = sor.wants;
+      uint takerGives = sor.gives;
+      uint offerId = sor.offerId;
+      MgvStructs.OfferPacked offer = sor.offer;
+      MgvStructs.OfferDetailPacked offerDetail = sor.offerDetail;
 
-        /* Keep cached copy of current `sor` values to restore them later to send to posthook. */
-        uint takerWants = sor.wants;
-        uint takerGives = sor.gives;
-        uint offerId = sor.offerId;
-        MgvStructs.OfferPacked offer = sor.offer;
-        MgvStructs.OfferDetailPacked offerDetail = sor.offerDetail;
+      /* If execution was successful, we update fillVolume downwards. Assume `mor.fillWants`: it is known statically that `mor.fillVolume - sor.wants` does not underflow. See the [`execute` function](#MgvOfferTaking/computeVolume) for details. */
+      if (mgvData == "mgv/tradeSuccess") {
+        mor.fillVolume -= mor.fillWants ? sor.wants : sor.gives;
+      }
 
-        /* If execution was successful, we update fillVolume downwards. Assume `mor.fillWants`: it is known statically that `mor.fillVolume - sor.wants` does not underflow. See the [`execute` function](#MgvOfferTaking/computeVolume) for details. */
-        if (mgvData == "mgv/tradeSuccess") {
-          mor.fillVolume -= mor.fillWants ? sor.wants : sor.gives;
-        }
+      /* We move `sor` to the next offer. Note that the current state is inconsistent, since we have not yet updated `sor.offerDetails`. */
+      /* It is known statically that `mor.initialGives - mor.totalGave` does not underflow since
+        1. `mor.totalGave` was increased by `sor.gives` during `execute`,
+        2. `sor.gives` was at most `mor.initialGives - mor.totalGave` from earlier step,
+        3. `sor.gives` may have been clamped _down_ during `execute` (to "`offer.wants`" if the offer is entirely consumed, or to `makerWouldWant`, cf. code of `execute`).
+      */
+      (sor.offerId, sor.local) = getNextBest(pair, mor, sor.offer, sor.local);
 
-        /* If an execution was attempted, we move `sor` to the next offer. Note that the current state is inconsistent, since we have not yet updated `sor.offerDetails`. */
-        if (mgvData != "mgv/notExecuted") {
-          /* It is known statically that `mor.initialGives - mor.totalGave` does not underflow since
-           1. `mor.totalGave` was increased by `sor.gives` during `execute`,
-           2. `sor.gives` was at most `mor.initialGives - mor.totalGave` from earlier step,
-           3. `sor.gives` may have been clamped _down_ during `execute` (to "`offer.wants`" if the offer is entirely consumed, or to `makerWouldWant`, cf. code of `execute`).
-        */
-          (sor.offerId, sor.local) = getNextBest(pair, mor, sor.offer, sor.local);
+      sor.offer = pair.offerData[sor.offerId].offer;
+      mor.gasreqLeft -= gasused;
```

Comment:
Here we haven't used all of the `offer_gasbase` gas set aside in line 209 since we're not done processing the offer yet. So I suspect this is an under approximation.

But maybe this is compensated for in line 156 where we set `offer_gasbase` aside for ending the order?

Reply:
You're on the right track. The reservation in 156 is needed because 258 is an underapproximation. It goes like this when taking two offers:
1. Line 258 reached first time, we have at least `offer_gasbase+gasreq1-gasused1` gas left for rounding off the market order due to what is set aside in 156 and the accounting in internalMarketOrder. It is larger due to line 209 also conditioning on additional offer_gasbase.
2. Line 258 reached the second time we have gasreqLeft be `offer_gasbase +gasreq1-gasused1+gasreq2-gasused2`, however, note that we will not execute offer2 unless gasleft() can also cover an additional offer_gasbase when reaching line 209. 

So we will never start executing an offer unless we also have `offer_gasbase` left on top of the `gasreq` . This should always be sufficient to complete a market order of 1 offer. For many offers this code relies on offer_gasbase being able to cover for the overhead of unwrapping the stack and doing the necessary updates.
Alternatively we should add some constant overhead for each offer on line 258 or let the reservation on line 156 be another value which we know is big enough to cover the unwrap of stack.

Note: gasreq for one offer is the worst case for that offer, so with a set of offers each recursion will set aside too much in case worst case is not hit, so on average a bit extra is set aside for each offer.

Reply:
Later when we add comments we should remember to include all this information.

Reply:
@lnist line numbers have gone stale!(?) I could use a refresh 😅

Reply:
Updated in the following:

You're on the right track. The reservation in 156 is needed because 260 is an underapproximation. It goes like this when taking two offers:
1. Line 260 reached first time, we have at least `offer_gasbase+gasreq1-gasused1` gas left for rounding off the market order due to what is set aside in 156 and the accounting in internalMarketOrder. It is larger due to line 211 also conditioning on additional offer_gasbase.
2. Line 260 reached the second time we have gasreqLeft be `offer_gasbase +gasreq1-gasused1+gasreq2-gasused2`, however, note that we will not execute offer2 unless gasleft() can also cover an additional offer_gasbase when reaching line 211. 

So we will never start executing an offer unless we also have `offer_gasbase` left on top of the `gasreq` . This should always be sufficient to complete a market order of 1 offer. For many offers this code relies on offer_gasbase being able to cover for the overhead of unwrapping the stack and doing the necessary updates.
Alternatively we should add some constant overhead for each offer on line 260 or let the reservation on line 156 be another value which we know is big enough to cover the unwrap of stack.

Note: gasreq for one offer is the worst case for that offer, so with a set of offers each recursion will set aside too much in case worst case is not hit, so on average a bit extra is set aside for each offer.

Reply:
Resolving obsolete discussion.